index.d.ts

```diff
@@ -0,0 +1,161 @@
+	 */
+	carryoverConcurrencyCount?: boolean;
+}
+
+interface DefaultAddOptions {
+	/**
+	 * Priority of operation. Operations with greater priority will be scheduled first.
+	 *
+	 * @default 0
+	 */
+	priority?: number;
+}
+
+type Task<TaskResultType> =
+	| (() => PromiseLike<TaskResultType>)
+	| (() => TaskResultType);
+
+/**
+ * Promise queue with concurrency control.
+ */
+export default class PQueue<
+	EnqueueOptionsType extends QueueAddOptions = DefaultAddOptions
+> extends EventEmitter {
+	/**
+	 * Size of the queue.
+	 */
+	readonly size: number;
+
+	/**
+	 * Number of pending promises.
+	 */
+	readonly pending: number;
+
+	/**
+	 * Whether the queue is currently paused.
+	 */
+	readonly isPaused: boolean;
+
+	constructor(options?: Options<EnqueueOptionsType>);
+
+	/**
+	 * Adds a sync or async task to the queue. Always returns a promise.
+	 *
+	 * @param fn - Promise-returning/async function.
+	 */
+	add<TaskResultType>(
+		fn: Task<TaskResultType>,
+		options?: EnqueueOptionsType
+	): Promise<TaskResultType>;
+
+	/**
+	 * Same as `.add()`, but accepts an array of sync or async functions.
+	 *
+	 * @param fn - Array of Promise-returning/async functions.
+	 * @returns A promise that resolves when all functions are resolved.
+	 */
+	addAll<TaskResultsType>(
+		fns: Task<TaskResultsType>[],
+		options?: EnqueueOptionsType
+	): Promise<TaskResultsType[]>;
+
+	/**
+	 * Can be called multiple times. Useful if you for example add additional items at a later time.
+	 *
+	 * @returns A promise that settles when the queue becomes empty.
+	 */
+	onEmpty(): Promise<void>;
+
+	/**
+	 * The difference with `.onEmpty` is that `.onIdle` guarantees that all work from the queue has finished. `.onEmpty` merely signals that the queue is empty, but it could mean that some promises haven't completed yet.
+	 *
+	 * @returns A promise that settles when the queue becomes empty, and all promises have completed; `queue.size === 0 && queue.pending === 0`.
+	 */
+	onIdle(): Promise<void>;
+
+	/**
+	 * Start (or resume) executing enqueued tasks within concurrency limit. No need to call this if queue is not paused (via `options.autoStart = false` or by `.pause()` method.)
+	 */
+	start(): void;
+
+	/**
+	 * Clear the queue.
+	 */
+	clear(): void;
+
+	/**
+	 * Put queue execution on hold.
+	 */
+	pause(): void;
+
+	addListener(event: 'active', listener: () => void): this;
+	on(event: 'active', listener: () => void): this;
+	once(event: 'active', listener: () => void): this;
+	prependListener(event: 'active', listener: () => void): this;
+	prependOnceListener(event: 'active', listener: () => void): this;
+	removeListener(event: 'active', listener: () => void): this;
+	off(event: 'active', listener: () => void): this;
+	removeAllListeners(event?: 'active'): this;
+	listeners(event: 'active'): (() => void)[];
+	rawListeners(event: 'active'): (() => void)[];
+	emit(event: 'active'): boolean;
+	eventNames(): 'active'[];
+	listenerCount(type: 'active'): number;
```

Comment:
This is not going to scale when we add more events, which is planned.

Reply:
Why not? We can add more events to the list. Or is it too cumbersome?

The problem is that node `EventEmitter` doesn't have any type variables for this. This is why we have to do this via method signature overrides. Of course it's still possible not to do it at all, but even node types use this pattern.

Reply:
It just feels super verbose. Is there really no better way of doing this that doesn't require duplicating all this for each event?

Reply:
I could put the list of events into a separate type but as soon events will start to emit different value types this won't be of much use any more.

As far as I know, this is the way to extend `EventEmitter`s but I would be glad if you could show me a better way to do it. I don't like it either.

Reply:
I don't really know a better way. Just feels weird that the status quo is not better than this. I guess to do something better, all of the types for EventEmitter would have had to be reworked. Example: https://github.com/sindresorhus/emittery#typescript

Reply:
The people at DefinitelyTyped are currently working on improving this status quo for newer TS versions.

Until TS 3.1 introduced the possibility to have separate versions of types for different versions of TS, there was no possibility to change node types without breaking thousands of packages. So I hope this situation will change in the near future.